native_client/ctcdecode/ctc_beam_search_decoder.cpp

```diff
@@ -205,11 +239,13 @@ DecoderState::decode(size_t num_results) const
   std::vector<Output> outputs;
   outputs.reserve(num_returned);
 
-  for (size_t i = 0; i < num_returned; ++i) {
+  for (PathTrie* prefix : prefixes_copy) {
     Output output;
-    prefixes_copy[i]->get_path_vec(output.tokens, output.timesteps);
-    output.confidence = scores[prefixes_copy[i]];
+    prefix->get_path_vec(output.tokens);
+    output.timesteps  = get_history(prefix->timesteps);
+    output.confidence = scores[prefix];
     outputs.push_back(output);
+    if (outputs.size() >= num_returned) break;
```

Comment:
I believe the opposite is true. 

The range-based for loop has been introduced to remove code like `prefixes_copy[i]`. Such code is a major cause of crashes, so every time I read this I have to wonder "Can this code crash ?". This makes it less readable. 
Then, I go back to the loop statement and read that `i` can go up to `num_returned-1`. I don't know, just by reading this line, whether this value is less than `prefixes_copy.size()`. So, I have to go go back again to check that. Again less readable.
In addition, it is possible that someone changes the definition of `num_returned` and it may cause a crash later in the loop.

In contrast, range-based for loops almost never crash. People used to complain about C/C++ segfaulting too often. Modern C++ has removed most causes of segfaults, and that's why ranged based for loops are almost always preferred.

Regarding the condition + break, it makes it clear that `outputs` is limited to a maximum size of `num_returned`, which is what the name "num_returned" means. With previous code, it was slightly less obvious.

Reply:
In C++ land you should be thinking "can this code crash?" all the time, regardless of using array indices or not. As evidenced by the last few commits in this PR. Modern C++ is a thin veil over the same crashy stuff. If someone changes the `num_returned` logic in your version, the code won't crash, but it'll silently do the wrong thing, so you need to be equally vigilant. In any case, I don't care enough about this case to force it, but watch out for buying more modernness than "Modern C++" can actually offer.

Reply:
We had quite a few crashes indeed ! 
It was due to null pointers, and thread safety, which are not covered well by "Modern C++" (smart pointers and thread_local help, but they don't prevent accessing bad memory).

Anyway, you have confirmed that you prefer like it was before, and this change is unrelated to this PR, so I should revert it.